### PR TITLE
Fix imports for `moduleResolution: nodenext` TS users (v4)

### LINKED ISSUE #6730: V4 broken for TS `nodenext` users
Lots of errors like this:
```
node_modules/@apollo/server/dist/esm/externalTypes/plugins.d.ts:6:344 - error TS2835: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean './requestPipeline.js'?
```

Repro: https://codesandbox.io/s/musing-albattani-s8v12i

Looks like we just need extensions _everywhere_, not just runtime imports.

### PR DESCRIPTION
**Pull request by trevor-scheer**
When installing `@apollo/server` v4 in a TS project using `"moduleResolution": "nodenext"`, TS throws a bunch of errors about how Apollo Server's imports need to use extensions. Specifically, we omitted extensions from all type-only imports since they didn't seem to be needed, but it's not hurting anything to add them and some use cases require it.

Fixes #6730

### CODE REVIEW

**Comment by trevor-scheer** [comment]
I'm able to reliably reproduce and then resolve the issue via this codesandbox:

https://codesandbox.io/s/hungry-bird-mzs69d

There seems to be some package caching, so it needs a thorough clean to reproduce.
1. install `@apollo/server@4.0.0-alpha.1`
2. if builds don't fail with above mentioned errors, delete `dist`, `yarn.lock`, and restart container
3. rebuild should show TS errors
4. install @apollo/server codesandbox build from this PR via: https://pkg.csb.dev/apollographql/apollo-server/commit/a61bb855/@apollo/server
5. TS errors are resolved

**Comment by glasser** [comment]
can you smoke-test it? (and ideally eslint but ISTR that wasn't possible)

**Comment by trevor-scheer** [comment]
@glasser I tried to reproduce this locally by cherry picking the new smoke test [fcb0295](https://github.com/apollographql/apollo-server/pull/6731/commits/fcb0295d5dc770dbf04a09de2d76f2971d888beb) commit onto `version-4` - test passes. So...the test I have so far is missing something that caused the original problem.

Needs more investigation, and possibly a more complex configuration than our current smoke tests use. i.e. I think I need a `"module": ...` or `"type": "module"` field in `package.json`...node needs to know it's using ESM.

**Comment by trevor-scheer** [comment]
My previous comment was correct - the package.json `"type": "module"` was the crux of this reproduction. To the point where the script running `tsc` actually has to be _in_ the correct directory (adjacent to the relevant package.json) for this test to work correctly!

File: .changeset/old-papayas-switch.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@apollo/server": patch
+---
+
+Use extensions for all imports to accommodate TS nodenext users
```
Comment (glasser) [inline .changeset/old-papayas-switch.md:5]: can you be explicit about whether this is module:nodenext, moduleResolution:nodenext, or both?

File: smoke-test/nodenext/package.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "nodenext-smoke-test",
+  "version": "0.0.0",
+  "type": "module",
+  "module": "./dist/smoke-test.js",
+  "dependencies": {
+    "@apollo/server": "../node_modules/@apollo/server",
```
Comment (glasser) [inline smoke-test/nodenext/package.json:7]: The rest of the smoketests install a built tarball — is that OK? or is this reading from inside smoke-test anyway i guess... huh.

**Review by glasser — APPROVED** [review]
It's a bummer that AFAIK we can't make eslint check this (since that's a lot faster than smoke-test). https://github.com/import-js/eslint-plugin-import/pull/2270 ...

File: smoke-test/nodenext/package.json (hunk 2 above)
Comment (trevor-scheer) [reply smoke-test/nodenext/package.json]: Simplified the reproduction a bit, and you're right that it's just looking up.
[cee95bb](https://github.com/apollographql/apollo-server/pull/6731/commits/cee95bbcc0034feb0253357aa3319923f2f5b0e6)

**Pull request merged**